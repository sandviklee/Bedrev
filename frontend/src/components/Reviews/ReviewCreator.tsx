import Button from "../Buttons/Button";
import * as AlertDialog from "@radix-ui/react-alert-dialog";
import * as Form from "@radix-ui/react-form";
import Field from "../Input/Field/Field";
import { FiXCircle } from "react-icons/fi";
import Rating from "@mui/material/Rating";
import styles from "./ReviewCreator.module.css";
import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { useMutation, useQuery, gql } from "@apollo/client";
import { ReviewJSON } from "../../types/types";
interface ReviewCreatorProps {
  workplace?: string;
  bedriftOrgNr: string;
}

const ReviewCreator = ({ workplace, bedriftOrgNr }: ReviewCreatorProps) => {
  const maxTitleWords = 30;
  const maxDescriptionWords = 100;
  const [title, setTitle] = useState<string>("");
  const [description, setDescription] = useState<string>("");
  const [rating, setRating] = useState<1 | 2 | 3 | 4 | 5>(1);
  const [userId, setUserId] = useState<number>(0);
  const [disableButton, setDisableButton] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [maxTitle, setMaxTitle] = useState<number>(maxTitleWords);
  const [maxDescription, setMaxDescription] =
    useState<number>(maxDescriptionWords);
  const authToken = localStorage.getItem("auth-token");
  const navigate = useNavigate();

  const USER_QUERY = gql`
    query {
      me {
        id
        reviews {
          postedBy {
            reviews {
              bedrift {
                organisasjonsnummer
              }
            }
          }
        }
      }
    }
  `;

  /**
   * Fetches data given above from the user query.
   * Sends the header with authorization token.
   */
  const { data, loading, error } = useQuery(USER_QUERY, {
    fetchPolicy: "cache-first", // Used for first execution
    nextFetchPolicy: "cache-first", // Used for subsequent executions
    context: { headers: { authorization: `Bearer ${authToken}` } },
  });

  /**
   * Checks if the user has written any
   * reviews for this company. If so,
   * it should disable the button.
   */
  useEffect(() => {
    if (data !== undefined) {
      setUserId(data["me"]["id"]);
      const map = data["me"]["reviews"].map((review: ReviewJSON) =>
        review["postedBy"]["reviews"].map(
          (companyReview: ReviewJSON) =>
            companyReview["bedrift"]["organisasjonsnummer"] ==
            parseInt(bedriftOrgNr),
        ),
      );
      if (map.length > 0) {
        setDisableButton(map[0].includes(true));
      }
    }
  }, [data]);

  /**
   * Updates the max title for each field.
   */
  useEffect(() => {
    setMaxTitle(maxTitleWords - title.length);
    setMaxDescription(maxDescriptionWords - description.length);
  }, [title, description]);

  const REVIEW_MUTATION = gql`
        mutation {
            createReview(
                title: "${title}"
                description: "${description}"
                rating: ${rating}
                bedriftOrgNr:${parseInt(bedriftOrgNr)}
                postedById: ${userId}
            ) {
                review {
                    id
                }
            }
        }
    `;

  const [createReview] = useMutation(REVIEW_MUTATION, {
    onCompleted: () => {
      setErrorMessage("");
      resetFields();
      navigate(0);
    },
    onError: (error) => {
      setErrorMessage(error.message);
    },
  });

  const resetFields = () => {
    setTitle("");
    setDescription("");
    setRating(1);
  };

  if (loading) return <p>Loading...</p>;
  if (error) return <pre>{error.message}</pre>;

  return (
    <>
      <AlertDialog.Root>
        <AlertDialog.Trigger asChild>
          <Button
            text={"Skriv en tilbakemelding"}
            flex
            disabled={disableButton}
            icon="Edit"
            px="[10px]"
          />
        </AlertDialog.Trigger>
        <AlertDialog.Portal>
          <AlertDialog.Overlay className={styles.AlertDialogOverlay} />
          <AlertDialog.Content className={styles.AlertDialogContent}>
            <AlertDialog.Title>
              <div className="flex flex-row w-full items-center">
                <p className="w-[1400px] sm:font-medium sm:text-base text-sm">
                  Skriv en tilbakemelding til {workplace}
                </p>
                <div className=" justify-end flex">
                  <AlertDialog.Cancel asChild>
                    <button
                      className="text-xl text-purple px-2 py-1"
                      onClick={() => resetFields()}
                    >
                      <FiXCircle />
                    </button>
                  </AlertDialog.Cancel>
                </div>
              </div>
            </AlertDialog.Title>

            <Form.Root
              className="w-full mt-[20px] flex flex-col gap-4"
              onSubmit={(event) => {
                event.preventDefault();
              }}
            >
              <Field
                name="Tittel"
                placeholder="Skriv en tittel her..."
                type="text"
                maxnumber={maxTitle}
                direction="col"
                label="Tittel"
                required
                border
                value={title}
                setValue={setTitle}
              />
              <Field
                name="Tilbakemelding"
                placeholder="Skriv en tilbakemelding her..."
                type="textarea"
                maxnumber={maxDescription}
                border
                direction="col"
                label="Tilbakemelding"
                required
                value={description}
                setValue={setDescription}
              />
            </Form.Root>
            {errorMessage !== "" && (
              <p className="text-base text-red">{errorMessage}</p>
            )}
            <div className="absolute">
              <p className="font-medium sm:py-[5px] py-[5px]">
                Gi en vurdering
              </p>
              <Rating
                name="simple-controlled"
                value={rating}
                onChange={(_, newValue) => {
                  setRating(newValue as 1 | 2 | 3 | 4 | 5);
                }}
              />
            </div>

            <div className="flex justify-end pt-[4%]">
              <Button
                text="Send inn"
                icon="ArrowRight"
                flex
                px="[4%]"
                onClick={() => createReview()}
              />
            </div>
          </AlertDialog.Content>
        </AlertDialog.Portal>
      </AlertDialog.Root>
    </>
  );
};

export default ReviewCreator;
