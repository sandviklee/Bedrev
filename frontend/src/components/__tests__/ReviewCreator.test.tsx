import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { BrowserRouter as Router } from "react-router-dom";
import { gql } from "@apollo/client";
import ReviewCreator from "../Reviews/ReviewCreator.tsx";

const USER_QUERY = gql`
  query {
    me {
      id
      reviews {
        postedBy {
          reviews {
            bedrift {
              organisasjonsnummer
            }
          }
        }
      }
    }
  }
`;

const REVIEW_MUTATION = gql`
  mutation {
    createReview(
      title: "Test title"
      description: "Test description"
      rating: 1
      bedriftOrgNr: 123456789
      postedById: 1
    ) {
      review {
        id
      }
    }
  }
`;

const mocks = [
  {
    request: {
      query: USER_QUERY,
    },
    result: {
      data: {
        me: {
          id: 1,
          reviews: [],
        },
      },
    },
  },
  {
    request: {
      query: REVIEW_MUTATION,
    },
    result: {
      data: {
        createReview: {
          review: {
            id: 1,
          },
        },
      },
    },
  },
];

describe("ReviewCreator component", () => {
  test("renders correctly", async () => {
    render(
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <ReviewCreator bedriftOrgNr="123456789" />
        </MockedProvider>
      </Router>,
    );

    expect(
      await screen.findByText("Skriv en tilbakemelding"),
    ).toBeInTheDocument();
    // click button to open dialog
    screen.getByText("Skriv en tilbakemelding").click();

    // check if fields and stuff are there
    expect(
      await screen.findByPlaceholderText("Skriv en tittel her..."),
    ).toBeInTheDocument();
    expect(
      await screen.findByPlaceholderText("Skriv en tilbakemelding her..."),
    ).toBeInTheDocument();
    expect(await screen.findByText("Gi en vurdering")).toBeInTheDocument();
    expect(await screen.findByText("Send inn")).toBeInTheDocument();
  });

  test("matches snapshot", async () => {
    const { asFragment } = render(
      <Router>
        <MockedProvider mocks={mocks} addTypename={false}>
          <ReviewCreator bedriftOrgNr="123456789" />
        </MockedProvider>
      </Router>,
    );
    expect(
      await screen.findByText("Skriv en tilbakemelding"),
    ).toBeInTheDocument();
    screen.getByText("Skriv en tilbakemelding").click();
    expect(await asFragment()).toMatchSnapshot();
  });
});
