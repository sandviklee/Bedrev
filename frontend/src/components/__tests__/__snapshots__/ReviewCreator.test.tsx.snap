// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReviewCreator component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="bg-white text-black hover:bg-white hover:text-black rounded-sm border-black border-2 transition-all py-2 w-fit text-sm px-[10px] group transition hover:bg-primary border-solid"
    type="button"
  >
    <div
      class="flex flex-row gap-2 justify-center pt-1 h-full px-6 align-middle"
    >
      Skriv en tilbakemelding
      <svg
        class="text-lg group-hover:-rotate-90 transition-all delay-75 pb-1"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 19l7-7 3 3-7 7-3-3z"
        />
        <path
          d="M18 13l-1.5-7.5L2 2l3.5 14.5L13 18l5-5z"
        />
        <path
          d="M2 2l7.586 7.586"
        />
        <circle
          cx="11"
          cy="11"
          r="2"
        />
      </svg>
    </div>
  </button>
</DocumentFragment>
`;
